### ISSUE TITLE
Embeded fonts text garbled after PDF generated few times  

### ISSUE THREAD

**Issue by sapradhan** [dompdf/dompdf#2485]
Using dompdf in laravel using a custom font defined in a `@fontface`  

initially it was working fine 
![image](https://user-images.githubusercontent.com/5070211/125109881-a38cd780-e103-11eb-80d8-bba7458b87f5.png)

but after PDF was generated few times the text started getting garbled. (Note the other text is part of a background image)
![image](https://user-images.githubusercontent.com/5070211/125110002-cae3a480-e103-11eb-87dc-da07d59eb2d3.png)

This does not happen if I use generic font families like `serif` `sans-serif` and fonts like `Helvetica` `Times` (implicitly available in PDF?)
![image](https://user-images.githubusercontent.com/5070211/125111687-ec459000-e105-11eb-85fd-c0650cb018d1.png)

Once text starts getting garbled it happens not just for external fonts but even for `dejavu sans` and `dejavu serif` which is bundled
![image](https://user-images.githubusercontent.com/5070211/125111558-c4eec300-e105-11eb-9622-47d066c81905.png)

so I am guessing this relates to font embedding? If I copy the garbled text and paste it in text editor, the text is correct. 

This happened to me on two different environments. First on a local development linux machine with `php artisan serve` and later on a shared cpanel hosted environment. Both are on php 7.4, laravel 8. I was using dompdf 0.8.6 earlier but have since then upgraded to 1.0.2 without any luck.

I have tried deleting font metrics, `.ufm.php` and other files generated in fonts dir but that does not help either.

**Comment by bsweeney** [comment]
Is this on isolated successive runs or during a single run that processes a bunch of documents?

**Comment by sapradhan** [comment]
> Is this on isolated successive runs or during a single run that processes a bunch of documents?

This is on isolated successive run. (if I understand it correctly) 

This is the code that generates the PDF, it is one PDF per request

        $data = ['salutation' => $salutation, 'user_name' => Auth::user()->name, 'todaydate' => $todayDate, 'user_id'=> $user_id];
        $pdf = App::make('dompdf.wrapper');
        $certificate = $pdf->loadView('certificate', $data)->setPaper('a4', 'landscape');

On both environments I have been unable to restore to correct behavior by any means once garbled text starts appearing.

**Comment by bsweeney** [comment]
By "isolated successive runs" I mean generating a bunch of PDFs in a single execuction of a PHP script. Something like:

```
foreach ($users as $user)
{
    // generate pdf
}
```

I wonder if the font cache is getting corrupted due to font subsetting. If you clear your font cache (default is the dompdf/lib/fonts directory) does the issue clear up? What about if you disable font subsetting (`$dompdf->set_option('isFontSubsettingEnabled', true);`)?

**Comment by sapradhan** [comment]
>> By "isolated successive runs" I mean generating a bunch of PDFs in a single execuction of a PHP script

No this is a single PDF generated per request. 

This is in a laravel project and I have published config - `dompdf.php` reads  

        /**
         * Whether to enable font subsetting or not.
         */
        "enable_font_subsetting" => false,

        /**
         * The location of the DOMPDF font cache directory
         *
         * This directory contains the cached font metrics for the fonts used by DOMPDF.
         * This directory can be the same as DOMPDF_FONT_DIR
         *
         * Note: This directory must exist and be writable by the webserver process.
         */
        "font_cache" => storage_path('fonts/'),


the content of the font cache folder looks like 
![image](https://user-images.githubusercontent.com/5070211/125116016-ce7b2980-e10b-11eb-9dc9-25d943ab80cf.png)

I have deleted these files (except for .ttf) and it does not help (I still get garbled text). Do I need to restart anything after I delete them?

**Comment by sapradhan** [comment]
I set font subsetting to true

        /**
         * Whether to enable font subsetting or not.
         */
        "enable_font_subsetting" => true,

and get this error 

![image](https://user-images.githubusercontent.com/5070211/125116611-b2c45300-e10c-11eb-8dba-2989255f6984.png)

**Comment by bsweeney** [comment]
Sorry, missed that you had already tried deleting the font metrics cache in the original post.

Are you always generating the PDF with the same text ("Mr. Student") while you're testing? Does the text start getting munged after the same number of renders or is it random? Can you post one of the garbled PDFs and the HTML+CSS?

**Comment by bsweeney** [comment]
You shouldn't need to restart anything after clearing the cache files. Dompdf should regenerate them if they don't exist (so that's something to check, that the files regenerate).

Do you have opcache enabled? Maybe PHP is caching some opcode from the metrics and that's causing the issue?

**Comment by bsweeney** [comment]
I wonder if this is similar to what's going on in #2438, for which I was unable to find a resolution as of yet. If you save the PDF to disk does the PDF appear correctly?

**Comment by sapradhan** [comment]
>Dompdf should regenerate them if they don't exist (so that's something to check, that the files regenerate).

Yes they are regenerated 

>Do you have opcache enabled? Maybe PHP is caching some opcode from the metrics and that's causing the issue?

I am not sure about this, I have not done anything to explicitly to enable it. This is a laravel project on a local dev machine and after installing dependencies using composer I ran `php artisan run`

>Are you always generating the PDF with the same text ("Mr. Student") while you're testing? 

No I have tried different texts. The characters are different if I change text but its munged up. I have also noticed different characters appear for different fonts even for same text.  
 
>Does the text start getting munged after the same number of renders or is it random? 

On local I think I got munged text from very beginning. On shared cpanel hosting I was able to generate a few, 3-4 PDFs before it started getting munged up. I may have changed a few files but not related to this.  

>And you have been unable to clear up the issue? 

No I haven't been able to get things working after it got munged. **One thing I tried was switching to GD renderer and font issue was fixed, however it generated a PNG instead of PDF** 

>I wonder if this is similar to what's going on in #2438, for which I was unable to find a resolution as of yet. 

May be. But there it seems character height is not calculated properly but the character itself is correct. In my case characters are messed up. 

>If you save the PDF to disk does the PDF appear correctly?

I havent tried this, will try and let you know. 

>Can you post one of the garbled PDFs and the HTML+CSS?

The only properly generated pdf I have has client logo in it so will revoke access once you can download 
[pdfs](https://drive.google.com/drive/folders/1ghUiOPmKja6iqQPz4DsWlqw4m8tMUD73?usp=sharing)
The munged up pdf is larger in size. And google drive pdf preview actually shows correct text but in wrong font. 

Here is the HTML + CSS,

```
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8" />
    <style type="text/css">

        * {
            margin: 0;
            padding: 0;
        }
        @font-face {
            font-family: 'Trajan Pro Regular';
            src: url("{{ storage_path('fonts/TrajanPro.ttf') }}");
            font-style: normal;
        }
        @font-face {
            font-family: 'Rubik Light';
            src: url("{{ storage_path('fonts/Rubik-Light.ttf') }}");
            font-style: thin;
        }

        .content_wrapper {
            height: 204mm;
            width: 297mm;
            background: url("{{ public_path('images/certificate-background1.png') }}");
            background-size: cover;
            background-repeat: no-repeat;
        }

        .name {
            font-size: .44in;
            line-height: .44in;
            font-family: 'trajan pro regular';
            /* font-family: 'sans-serif'; */
            text-align: center;
            padding-top: 4.8in;
            /* padding-top: 4.9in; */
            color: #006CB4;
        }

        .date {
            padding-top: 0.3in;
            /* padding-top: 0.37in; */
            padding-left: 6.85in;
            /* font-family: 'sans-serif'; */
            font-family: 'rubik light';
            font-size: 1.1em;
        }

        .id {
            padding-top: 1.8in;
            padding-right: .4in;
            text-align: right;
            font-family: 'sans-serif';
        }
    </style>
</head>

<body>
    <div class="content_wrapper">

        <div class="name">Mr John Doe</div>

        <div class="date">July 9, 20201</div>

        <div class="id">ID: 2</div>

    </div>
</body>

</html>
```

and PHP code 

        $data = ['salutation' => $salutation, 'user_name' => Auth::user()->name, 'todaydate' => $todayDate, 'user_id'=> $user_id];
        $pdf = App::make('dompdf.wrapper');
        $certificate = $pdf->loadView('certificate', $data)->setPaper('a4', 'landscape');

composer requires 

        "barryvdh/laravel-dompdf": "^0.9.0",


config `dompdf.php`
```

<?php

return array(

    'show_warnings' => false,   // Throw an Exception on warnings from dompdf
    'orientation' => 'portrait',
    'defines' => array(

        "font_dir" => storage_path('fonts/'), // advised by dompdf (https://github.com/dompdf/dompdf/pull/782)

        "font_cache" => storage_path('fonts/'),

        "temp_dir" => sys_get_temp_dir(),

        "chroot" => realpath(base_path()),

        "enable_font_subsetting" => false,

        "pdf_backend" => "CPDF",

        //"DOMPDF_PDFLIB_LICENSE" => "your license key here",

        "default_media_type" => "screen",

        "default_paper_size" => "a4",

        "default_font" => "serif",

        "dpi" => 96,

        "enable_php" => false,

        "enable_javascript" => true,

        "enable_remote" => true,


        "font_height_ratio" => 1.1,

        "enable_html5_parser" => false,
    ),


);
```

**Comment by bsweeney** [comment]
If you look farther down in the 2438 discussion you can see that text in the linked streamed PDF is being corrupted, not just the line heights. Your code seems fine, there must be something more basic I'm missing.

What is the underlying web server, OS, hardware architecture, etc on the systems (if you know for your host)? Anything that could be running in the PDF generation/delivery pipeline.

Is your HTTP server running HTTP/2?

Have you tried capturing the output and streaming it manually instead of using the `stream` method? Something like:

```
$pdf = $dompdf->output();
header('Content-Disposition: attachment; filename="render.pdf"');
header('Content-Type: application/pdf');
echo $pdf;
exit();
```

**Comment by sapradhan** [comment]
>    If you save the PDF to disk does the PDF appear correctly?

Thank you for pointing this out, PDF generated this way was good. 

And I figured out what was corrupting my output. 

```
$certificate = $pdf->loadView('certificate', $data)->setPaper('a4', 'landscape');

// conditional code that sends mail with PDF as attachment using $certificate->output()

return $certificate->download()
```
Apparently calling `$certificate->output()` corrupts the pdf when it is later sent to download. I saved the PDF as a temp file and used to send email and serve off the file and it works now. 

Really appreciate your prompt responses and for taking time to guide me through this. 
Should we close this issue ?
